Delegate template parameter incorrectly treated as compile-time

This code:

```c#
using Metalama.Framework.Aspects;
using Metalama.Framework.Code;

class MyAspect : TypeAspect
{
    public override void BuildAspect(IAspectBuilder<INamedType> builder)
    {
        builder.Advice.IntroduceMethod(builder.Target, nameof(Template));
    }

    [Template]
    void Template(Func<object> f)
    {
        f();
    }
}

[MyAspect]
class Target
{
}
```

fails when building the compile-time compilation with:

> error CS0149: Method name expected

A variation that uses `Action` instead fails with:

> error LAMA0200: A compile-time value of type 'void' was used in a context where a run-time value was expected.

## Phase 1: Understanding the Issue

**Issue:** When a template method has a parameter of a delegate type (`Func<T>`, `Action`, etc.), Metalama incorrectly classifies the parameter as compile-time instead of run-time.

**Expected behavior:** Delegate-type parameters in templates (without `[CompileTime]` annotation) should be treated as run-time parameters. They should:
- Remain in the generated method signature
- Be callable at run-time

**Actual behavior:** The delegate parameters are treated as compile-time, causing:
1. `Func<object> f` case: `error CS0149: Method name expected` — the delegate parameter is removed from compile-time code, making `f()` invalid
2. `Action` variation: `error LAMA0200` — the delegate invocation is wrongly treated as a compile-time expression returning `void`

**Root cause hypothesis:** The compile-time/run-time scope classification logic incorrectly identifies delegate types as compile-time-only, possibly because the classification doesn't properly handle delegate types that wrap run-time-or-compile-time generic arguments.

Moving to Phase 2: reproducing the bug with a regression test.

— Claude for @gfraiteur

## Phase 2: Bug Reproduced

The bug has been reproduced with a regression test in PR #1395.

**Test:** `IntroduceMethod_DelegateParameter` in `Metalama.Framework.Tests.AspectTests/Tests/Aspects/TemplateParameters/`

The test confirms that when a template method has delegate-type parameters (`Func<object>`, `Action`), the compile-time compilation fails with:
- `error CS0149: Method name expected` for `f()`
- `error CS0149: Method name expected` for `a()`

This confirms the delegate parameters are incorrectly removed from the compile-time code (treated as compile-time parameters instead of run-time parameters).

Moving to Phase 3: implementing the fix.

— Claude for @gfraiteur

--- a/Metalama.Framework/src/Metalama.Framework.Engine/Templating/TemplateAnnotator.cs
+++ b/Metalama.Framework/src/Metalama.Framework.Engine/Templating/TemplateAnnotator.cs
@@ -1247,7 +1247,12 @@ internal sealed partial class TemplateAnnotator : SafeSyntaxRewriter, IDiagnosti
                     SeparatedList( transformedArguments, node.ArgumentList.Arguments.GetSeparators() ),
                     node.ArgumentList.CloseParenToken ) );
 
-            updatedInvocation = updatedInvocation.AddScopeAnnotation( expressionScope );
+            // When the expression is a run-time template parameter (e.g. a delegate), the invocation produces
+            // a run-time value, not another template parameter reference. Use RunTimeOnly so the invocation
+            // gets transformed to syntax factory code.
+            var invocationScope = expressionScope == RunTimeTemplateParameter ? RunTimeOnly : expressionScope;
+
+            updatedInvocation = updatedInvocation.AddScopeAnnotation( invocationScope );
 
             if ( expressionScope == CompileTimeOnlyReturningRuntimeOnly && this._syntaxTreeAnnotationMap.GetExpressionType( node ) is
                     { TypeKind: TypeKind.Dynamic } )

--- a/Metalama.Framework/src/tests/Metalama.Framework.Tests.AspectTests/Tests/Aspects/TemplateParameters/IntroduceMethod_DelegateParameter.cs
+++ b/Metalama.Framework/src/tests/Metalama.Framework.Tests.AspectTests/Tests/Aspects/TemplateParameters/IntroduceMethod_DelegateParameter.cs
@@ -1,0 +1,37 @@
+// Copyright (c) 2020-2025 SharpCrafters s.r.o. and contributors.
+// SharpCrafters s.r.o. licenses this file to you under either the MIT license or a proprietary license, depending on the repository from which it was obtained.
+// Refer to LICENSE.md in the repository root for complete details.
+
+using System;
+using Metalama.Framework.Advising;
+using Metalama.Framework.Aspects;
+using Metalama.Framework.Code;
+
+namespace Metalama.Framework.Tests.AspectTests.Tests.Aspects.TemplateParameters.IntroduceMethod_DelegateParameter;
+
+public class MyAspect : TypeAspect
+{
+    public override void BuildAspect( IAspectBuilder<INamedType> builder )
+    {
+        builder.IntroduceMethod( nameof(FuncTemplate) );
+        builder.IntroduceMethod( nameof(ActionTemplate) );
+    }
+
+    [Template]
+    private void FuncTemplate( Func<object> f )
+    {
+        f();
+    }
+
+    [Template]
+    private void ActionTemplate( Action a )
+    {
+        a();
+    }
+}
+
+// <target>
+[MyAspect]
+internal class Target
+{
+}

--- a/Metalama.Framework/src/tests/Metalama.Framework.Tests.AspectTests/Tests/Aspects/TemplateParameters/IntroduceMethod_DelegateParameter.t.cs
+++ b/Metalama.Framework/src/tests/Metalama.Framework.Tests.AspectTests/Tests/Aspects/TemplateParameters/IntroduceMethod_DelegateParameter.t.cs
@@ -1,0 +1,13 @@
+[MyAspect]
+internal class Target
+{
+  private void FuncTemplate(global::System.Func<global::System.Object> f)
+  {
+    f();
+  }
+
+  private void ActionTemplate(global::System.Action a)
+  {
+    a();
+  }
+}

--- a/Metalama.Framework/src/tests/Metalama.Framework.Tests.AspectTests/Tests/Aspects/TemplateParameters/IntroduceMethod_DelegateParameter.t.cs
+++ b/Metalama.Framework/src/tests/Metalama.Framework.Tests.AspectTests/Tests/Aspects/TemplateParameters/IntroduceMethod_DelegateParameter.t.cs
@@ -1,13 +1,12 @@
 [MyAspect]
 internal class Target
 {
-  private void FuncTemplate(global::System.Func<global::System.Object> f)
-  {
-    f();
-  }
-
   private void ActionTemplate(global::System.Action a)
   {
     a();
+  }
+  private void FuncTemplate(global::System.Func<global::System.Object> f)
+  {
+    f();
   }
 }

